### PR TITLE
StringTake

### PR DESCRIPTION
**Pull request by mmatera**
I just added this functionality to the strings.py module

### CODE REVIEW

File: mathics/builtin/strings.py [hunk 1]
```diff
@@ -699,3 +699,109 @@ class StringQ(Test):
...
+        'StringTake[string_,n_]'
+        if not isinstance(string, String):
+            return evaluation.message('StringTake', 'strse', string)
+        if isinstance(n, Integer):
+            pos=n.value
+            if pos>0:
+                return String(string.get_string_value()[:pos])
+            if pos<0:
+                return String(string.get_string_value()[pos:])
+        return evaluation.message('StringTake', 'naioli', n)
```
Comment (sn6uv) [inline mathics/builtin/strings.py]: What's the expected behaviour for e.g. `StringTake ["abc", 0]` ?

File: mathics/builtin/strings.py [hunk 2]
```diff
@@ -699,3 +699,109 @@ class StringQ(Test):
...
+        if not isinstance(string, String):
+            return evaluation.message('StringTake', 'strse', string)
+        if isinstance(ni, Integer) and  isinstance(nf, Integer)  :
+            if ni.value==0 or nf.value==0:
+                return evaluation.message('StringTake', 'badrange', ni,nf)
+            fullstring=string.get_string_value()
+            posi=ni.value        
+            if posi<0:
+                posi=len(fullstring)+posi+1
+            posf=nf.value
```
Comment (sn6uv) [inline mathics/builtin/strings.py]: Is the aim to have `0 < posi < posf <= len(fullstring)` ? There might be some more edge cases to consider here. e.g. `ni=-10` and `fullstring="ab"`. The solution to this case should probably be to raise an error

File: mathics/builtin/strings.py [hunk 3]
```diff
@@ -699,3 +699,109 @@ class StringQ(Test):
...
+        if isinstance(ni, Integer) and  isinstance(nf, Integer)  :
+            if ni.value==0 or nf.value==0:
+                return evaluation.message('StringDrop', 'badrange', ni,nf)
+            fullstring=string.get_string_value()
+            posi=ni.value        
+            if posi<0:
+                posi=len(fullstring)+posi+1
+            posf=nf.value
+            if posf<0:
+                posf=len(fullstring)+posf+1
```
Comment (sn6uv) [inline mathics/builtin/strings.py]: same issue as StringTake

**Comment by sn6uv** [comment]
A couple of minor indexing issues to fix up. Looks good otherwise.

**Comment by mmatera** [comment]
I think that many of the previous issues was fixed. It still remains to implement the case 
StringTake[string,{ni,nf,inc}]
and
StringDrop[string,{ni,nf,inc}]

**Comment by sn6uv** [comment]
The functionality looks good. Would you mind cleaing up the code a little? I think some tabs got mixed into the source code. We generally try to stick to [PEP8](https://www.python.org/dev/peps/pep-0008/) in the coding style although I'm not super strict about this (especially line length < 80). Other than that I'd say it's good to merge.

File: mathics/builtin/strings.py [hunk 4]
```diff
@@ -699,3 +699,182 @@ class StringQ(Test):
...
+        posf = nf.value
+        if posf < 0:
+            posf = lenfullstring + posf + 1
+        if (posf > lenfullstring or posi > lenfullstring
+            or posf <= 0 or posi <= 0):
+            #positions out or range
+            return evaluation.message('StringDrop',
+                                      'drop', ni, nf, fullstring)
+        if posf < posi:
+            return string           # this is what actually mma does
```
Comment (sn6uv) [inline mathics/builtin/strings.py]: could you add a test for this behaviour

File: mathics/builtin/strings.py [hunk 5]
```diff
@@ -699,3 +699,182 @@ class StringQ(Test):
...
+        return String(fullstring[:(posi - 1)] + fullstring[posf:])
+
+
+    def apply3_(self, string, ni, evaluation):
+        'StringDrop[string_,{ni_Integer}]'
```
Comment (sn6uv) [inline mathics/builtin/strings.py]: Could you add this pattern to the docs as well please.

File: mathics/builtin/strings.py [hunk 6]
```diff
@@ -699,3 +699,182 @@ class StringQ(Test):
...
+        if (posf > lenfullstring or posi > lenfullstring or
+            posf <= 0 or posi <= 0):
+            #positions out of range
+            return evaluation.message('StringTake', 'take', ni, nf, fullstring)
+        if posf < posi:
+            String("")
+        return String(fullstring[(posi - 1):posf])
+
+    def apply3_(self, string, ni, evaluation):
+        'StringTake[string_,{ni_}]'
```
Comment (sn6uv) [inline mathics/builtin/strings.py]: could you add this pattern to the docs as well please

File: mathics/builtin/strings.py [hunk 7]
```diff
@@ -699,3 +699,182 @@ class StringQ(Test):
...
+        if pos > len(string.get_string_value()):
+            return evaluation.message('StringTake', 'take', 1, pos, string)
+        if pos < -len(string.get_string_value()):
+            return evaluation.message('StringTake', 'take', pos, -1, string)
+        if pos > 0:
+            return String(string.get_string_value()[:pos])
+        if pos < 0:
+            return String(string.get_string_value()[pos:])
+        if pos == 0:
+            return String("")             # it is what mma does
```
Comment (sn6uv) [inline mathics/builtin/strings.py]: probably a good idea to add a (hidden) test for this behaviour.

**Comment by sn6uv** [comment]
Looking good. Just a few things missing from the docs/tests. I'm happy to merge this once those issues have been fixed.

File: mathics/builtin/strings.py (hunk 4 above)
Comment (mmatera) [inline mathics/builtin/strings.py]: Here, what you mean is to test 
Drop["abcdef",{3,2}]
"abcdef"
?

File: mathics/builtin/strings.py (hunk 5 above)
Comment (mmatera) [inline mathics/builtin/strings.py]: Are you refering here to apply3? or the behaviour of apply2 when posf<posi ?

File: mathics/builtin/strings.py (hunk 7 above)
Comment (mmatera) [inline mathics/builtin/strings.py]: How can I do it?

Comment (wolfv) [inline mathics/builtin/strings.py]: Thats a test in the comments

    >> StringDrop["abcde", 2]
    = cde

And thats a hidden test:
    #> StringDrop["abcde", 2]
    = cde

I think (or #>>)

File: mathics/builtin/strings.py (hunk 4 above)
Comment (sn6uv) [inline mathics/builtin/strings.py]: exactly

**Pull request merged**